src/service.py

```diff
@@ -0,0 +1,143 @@
+UPSTREAM_SNAP = "golang-openstack-exporter"
+
+
+class SnapService:
+    """A class representing the snap service(s)."""
+
+    def __init__(self, client: snap.Snap) -> None:
+        """Initialize the class."""
+        self.snap_client = client
+
+    def is_active(self) -> bool:
+        """Return True if all snap service(s) is / are active."""
+        return all(service.get("active", False) for service in self.snap_client.services.values())
+
+    def restart_and_enable(self) -> None:
+        """Restart and enable the snap service.
+
+        Ensure:
+        - service is running
+        - service is enabled
+        - service is restarted if already running to apply updated config
+        """
+        # This is safe to always run,
+        # because restarting when service is disabled has no effect,
+        # and restarting when enabled but stopped has the same effect as start.
+        self.snap_client.restart()
+        # This is idempotent, so ok to always run to ensure it's started and enabled
+        self.snap_client.start(enable=True)
+
+    def stop(self) -> None:
+        """Stop and disable the snap service."""
+        self.snap_client.stop(disable=True)
+
+    def configure(self, snap_config: dict[str, Any]) -> None:
+        """Configure the snap service."""
+        # Bail out or it will crash on self.snap_client.set() method
+        if not snap_config:
+            logger.warning("empty snap config: %s, skipping...", snap_config)
+            return
+
+        self.snap_client.set(snap_config, typed=True)
+
+    @property
+    def present(self) -> bool:
+        """Check if the snap client is present or not."""
+        return self.snap_client.present
+
+
+def snap_install_or_refresh(resource: Optional[str], channel: str) -> None:
+    """Install or refresh the snap.
+
+    Before installing the snap, it will try to remove the upstream snap that could be installed on
+    older versions of this charm.
+
+    If the channel is changed, the snap.add method is able to identify and refresh the charm to the
+    new channel.
+
+    Raises an exception on error.
+    """
+    remove_upstream_snap()
+    try:
+        if resource:
+            logger.debug("installing %s from resource.", SNAP_NAME)
+            # installing from a resource if installed from snap store previously is not problematic
+            snap.install_local(resource, dangerous=True)
+        else:
+            # installing from snap store if previously installed from resource is problematic, so
+            # it's necessary to remove it first
+            remove_snap_as_resource()
+            logger.debug("installing %s from snapcraft store", SNAP_NAME)
+            snap.add(SNAP_NAME, channel=channel)
+    except snap.SnapError as e:
+        logger.error("failed to install snap: %s", str(e))
+        raise e
+    else:
+        logger.info("installed %s snap.", SNAP_NAME)
+        workaround_bug_268()
+
+
+def remove_upstream_snap() -> None:
```

Comment:
Yes, then the charm will be error status. It's an exception that charm can't not handle.

Reply:
Is this documented somewhere? Is there any way for the charm to recover automatically?

Reply:
After discussion, we couldn't find an efficient solution. Therefore, it seems reasonable to let the charm enter 'error' status. In this case, should we provide documentation on Discourse?

Reply:
It sounds reasonable to include that in the documentation yes and error state makes sense as well

Reply:
I've opened [Issue #108](https://github.com/canonical/openstack-exporter-operator/issues/108) on GitHub to track this. We'll update you here once it's resolved

Reply:
This function is for upgrading from revision <= 27 to revision >= 28.  We have a document at https://charmhub.io/openstack-exporter/docs/upgrade-from-27-or-older that details the manual steps required.  It doesn't explicitly say what cases could cause error state though - what kind of docs are you looking for on this? This could fail for a number of unexpected reasons, and it would be up to the user to debug it.

Reply:
I would say think about the user journey. They are upgrading from an older revision and just ran the upgrade command and it fails. How would they find out it fails? Perhaps from the juju status message? How would they know to go from there to the docs? In the docs, what are the different failure scenarios, and how would the user be able to recover from those?

Reply:
@jdkandersson ok thanks, so it seems that it's not a strong opinion on whether to go Blocked or Error; it's more about how do we have the best UX for the situation?

Reply:
I think on one of the other comments on this PR there is a long discussion around error versus blocked state, so there are string opinions on that in the charming space 🙂 the TL;DR version is blocked if there is a specific action the operator can take to resolve, error otherwise

And you are right, charm should also have a great UX

Reply:
Thanks for your comments. :)

Since https://github.com/canonical/openstack-exporter-operator/pull/110 , unsuccessful removal of the upstream snap will result in error status.

Reply:
ok great, looks fine